src/java.base/share/classes/java/io/ObjectStreamDefaultSupport.java

```diff
@@ -0,0 +1,136 @@
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+
+import jdk.internal.access.JavaObjectStreamDefaultSupportAccess;
+import jdk.internal.access.SharedSecrets;
+import jdk.internal.util.ByteArray;
+
+/**
+ *
+ */
+final class ObjectStreamDefaultSupport {
+
+    // todo: these could be constants
+    private static final MethodHandle DRO_HANDLE;
+    private static final MethodHandle DWO_HANDLE;
+
+    static {
+        try {
+            MethodHandles.Lookup lookup = MethodHandles.lookup();
+            MethodType droType = MethodType.methodType(void.class, ObjectStreamClass.class, Object.class, ObjectInputStream.class);
+            DRO_HANDLE = lookup.findStatic(ObjectStreamDefaultSupport.class, "defaultReadObject", droType);
+            MethodType dwoType = MethodType.methodType(void.class, ObjectStreamClass.class, Object.class, ObjectOutputStream.class);
+            DWO_HANDLE = lookup.findStatic(ObjectStreamDefaultSupport.class, "defaultWriteObject", dwoType);
+        } catch (NoSuchMethodException e) {
+            throw new NoSuchMethodError(e.getMessage());
+        } catch (IllegalAccessException e) {
+            throw new IllegalAccessError(e.getMessage());
+        }
+    }
+
+    private static void defaultReadObject(ObjectStreamClass streamClass, Object obj, ObjectInputStream ois) throws IOException, ClassNotFoundException {
+        ObjectInputStream.GetField getField = ois.readFields();
+        byte[] bytes = new byte[streamClass.getPrimDataSize()];
+        Object[] objs = new Object[streamClass.getNumObjFields()];
+        for (ObjectStreamField field : streamClass.getFields(false)) {
+            int offset = field.getOffset();
+            String fieldName = field.getName();
+            switch (field.getTypeCode()) {
+                case 'B' -> bytes[offset] = getField.get(fieldName, (byte) 0);
+                case 'C' -> ByteArray.setChar(bytes, offset, getField.get(fieldName, (char) 0));
+                case 'D' -> ByteArray.setDoubleRaw(bytes, offset, getField.get(fieldName, 0.0));
+                case 'F' -> ByteArray.setFloatRaw(bytes, offset, getField.get(fieldName, 0.0f));
+                case 'I' -> ByteArray.setInt(bytes, offset, getField.get(fieldName, 0));
+                case 'J' -> ByteArray.setLong(bytes, offset, getField.get(fieldName, 0L));
+                case 'S' -> ByteArray.setShort(bytes, offset, getField.get(fieldName, (short) 0));
+                case 'Z' -> ByteArray.setBoolean(bytes, offset, getField.get(fieldName, false));
+                case '[', 'L' -> objs[offset] = getField.get(fieldName, null);
+                default -> throw new IllegalStateException();
+            }
+        }
+        streamClass.setPrimFieldValues(obj, bytes);
+        streamClass.setObjFieldValues(obj, objs);
+    }
+
+    private static void defaultWriteObject(ObjectStreamClass streamClass, Object obj, ObjectOutputStream oos) throws IOException {
```

Comment:
Ditto here, describing a subclasses OOS can override putFields() to return a PutField instance that extracts the field values from the object and copies them into the PutField instance making the value available to be put into the stream in a custom format.